lib/rubocop/cop/rspec/mock_not_stub.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that stubs aren't added to mocks.
+      #
+      # @example
+      #
+      #   # bad
+      #   expect(foo).to receive(:bar).with(42).and_return("hello world")
+      #
+      #   # good (without spies)
+      #   allow(foo).to receive(:bar).with(42).and_return("hello world")
+      #   expect(foo).to receive(:bar).with(42)
+      #
+      #   # good (with spies)
+      #   allow(foo).to receive(:bar).with(42).and_return("hello world")
+      #   expect(foo).to have_received(:bar).with(42)
+      #
+      class MockNotStub < Cop
+        MSG = "Don't stub your mock.".freeze
+
+        def_node_matcher :message_expectation_with_return_argument, <<-PATTERN
+          (send
+            (send nil? :expect ...) :to
+            $(send #receive :and_return _)
+          )
+        PATTERN
+
+        def_node_matcher :message_expectation_with_return_block, <<-PATTERN
+          (send
+            (send nil? :expect ...) :to
+            $(block #receive (args) _)
+          )
+        PATTERN
+
+        def_node_matcher :receive, <<-PATTERN
+          {
+            (send nil? :receive ...)
+            (send (send nil? :receive ...) :with ...)
```

Comment:
Wondering if it will work to change this to `def_node_search` so that it supports arbitrary nesting.
Due to the wide range of allowed qualifiers (e.g. `once`, `at_least`, `ordered`, ...) it's near impossible to match them with a `def_node_matcher`.

I've faced a similar problem [here](https://github.com/rubocop-hq/rubocop-rspec/pull/726/files#diff-faa46202d8b79d8bf965118100cff4d2R73), and using `def_node_search` solved it.